### ISSUE TITLE
Infer's Cost Analysis documentation

### ISSUE THREAD

**Issue by yunjeong-lee** [facebook/infer#1389]
There are 3 things--2 comments and 1 question--in this issue.

1. First, there is a typo in the [cost analysis document](https://fbinfer.com/docs/checker-cost/).

The command has to be modified from:
* `run infer reportdiff --costs-current current-costs-report.json --costs-previous current-costs-report`.

to:
* `run infer reportdiff --costs-current current-costs-report.json --costs-previous previous-costs-report`.

2. Also, it would be helpful if the document says that I need to move these costs reports out of the infer-out directory and run the command.

3. Lastly, I have a question on the following example (which is provided in the same documentation)

- Infer v0.17.0
- Ubuntu 18.04.5

I ran Infer's cost analysis by the following sequence on the example provided by [Infer's documentation](https://fbinfer.com/docs/checker-cost/).

Here is a sequence of steps taken:

1. Created the following Java file `miniCost.java` and ran `infer --cost-only -- javac miniCost02.java` from the same directory:

```
import java.util.*;
class miniCost {
  public void loop(ArrayList<Integer> list){
    for (int i = 0; i <= list.size(); i++){
    }
  }
}
```

From the terminal, I got:
```
$ infer --cost-only -- javac miniCost01.java
Capturing in javac mode...
Found 1 source file to analyze in /home/yunjeonglee/Documents/program-repair/infer-runs/miniRuns/mini01/infer-out


Analysis finished in 971mss

  No issues found  

```

2. Then, renamed `costs-report.json` to `previous-costs-report.json`. (This file was automatically generated in `infer-out` directory.)

3. Modified the function in the java file as follows:

```
import java.util.*;
class miniCost {
  public void loop(ArrayList<Integer> list){
    for (int i = 0; i <= list.size(); i++){
      	foo(i); // newly added function call
    }
  }
  public int foo(int n){
    n += 1;
    return n;
  }
}
```

4. Re-ran infer's cost analysis on the same file, using the same command as Step 1 above.

```
infer --cost-only -- javac miniCost01.java
Capturing in javac mode...
Found 1 source file to analyze in /home/yunjeonglee/Documents/program-repair/infer-runs/miniRuns/mini01/infer-out


Analysis finished in 1.267ss

  No issues found  

```

5. Renamed `costs-report.json` to `current-costs-report.json`.
Moved these `previous-costs-report.json` and `current-costs-report.json` outside the `infer-out` directory to `mini01` directory.

6. Then, ran the following command `infer reportdiff --costs-current current-costs-report.json --costs-previous previous-costs-report` from the `mini01` directory.

7. Afterwards, when I checked the `differential/introduced`, I see that no complexity has been introduced.

Basically, inside the `differential` directory, I see the following files:
```
.../differential$ ls
costs_summary.json  fixed.json  introduced.json  preexisting.json
```
And `fixed.json`, `introduced.json`, and `preexisting.json` files all have just `[]`, meaning nothing is being reported here.

And when `costs_summary.json` had the following result:
```
$ cat differential/costs_summary.json 
{"top":{"current":1,"previous":1},"zero":{"current":3,"previous":2},"degrees":[{"degree":0,"current":2,"previous":1}]}
```
But, why is this the case? I clearly introduced some complexity by adding `foo` function above, right?

Is there any obvious way to introduce complexities? I want to be able to see some outcome in the `introduced.json` file.

**Comment by skcho** [comment]
Hello @yunjeong-lee.

> First, there is a typo in the cost analysis document.

Thank you for letting us know. 👍  We will fix it.

> Also, it would be helpful if the document says that I need to move these costs reports out of the infer-out directory and run the command.

Oh, rerunning of Infer is removing the renamed file. We will fix it too.

> But, why is this the case? I clearly introduced some complexity by adding foo function above, right?

The reason is that the cost checker reports an complexity increase issue only when its **degree** is increased.
https://github.com/facebook/infer/blob/7cc66992fa73ead62adc4e6ff8ea57d283f25a68/infer/src/integration/Differential.ml#L339
While `foo` was added in the example, overall complexity of `loop` is still `O(list.size())` since `foo`'s complexity is `O(1)`. If we add `foo` as

```
  public int foo(int n){
    for (int i = 0; i < n; i++) {}
    return n;
  }
```

We will get non-empty `introduced.json` as follows.

```
[{
  "bug_type":"EXECUTION_TIME_COMPLEXITY_INCREASE",
  "qualifier":"Time complexity of `loop` has **increased** from `O(list.length)` to `O(list.length × list.length)`. Please make sure this is an expected change.",
  "procedure":"miniCost.loop(java.util.ArrayList):void",
  "file":"miniCost.java",
  ...,
  "bug_type_hum":"Execution Time Complexity Increase"
}]
```

**Comment by jvillard** [comment]
Please re-open if you have more questions.
